workflow_imaging/ingest.py

```diff
@@ -1,68 +1,70 @@
 
-    Equipment.insert(scanners, skip_duplicates=True)
-    Session.insert(sessions)
-    scan.Scan.insert(scans)
-    print(f'Inserted {len(sessions)} session(s)')
+    for session in sessions_dict:
+        sess_dir = pathlib.Path(session['session_dir'])
+        tiff_filepaths = [fp.as_posix() for fp in sess_dir.glob('*.tif')]
 
-    # populate ScanInfo
-    print('---- Populate ScanInfo ----')
-    scan.ScanInfo.populate(suppress_errors=True, display_progress=True)
+        try:  # attempt to read .tif as a scanimage file
+            loaded_scan = scanreader.read_scan(tiff_filepaths)
+        except Exception as e:
+            print(f'ScanImage loading error: {tiff_filepaths}\n{str(e)}')
+            loaded_scan = None
+
+        if loaded_scan is not None:
+            recording_time = get_scanimage_acq_time(loaded_scan)
+            header = parse_scanimage_header(loaded_scan)
+            session_key = {'subject': session['subject'], 'session_datetime': recording_time}
+            scanner = header['SI_imagingSystem'].strip('\'')
+            if session_key not in Session.proj():
+                scanners.append({'scanner': scanner})
+                sessions.append(session_key)
+                scans.append({**session_key, 'scan_id': 0, 'scanner': scanner})
+                session_directories.append({**session_key, 'session_dir': session['session_dir']})
+                
+                parameter_set = [int(s) for s in session['paramset'].split(' ')]
+                for param in parameter_set:
+                    processing_tasks.append({**session_key, 'scan_id': 0, 'paramset_idx': param, 'task_mode': 'load'})
+
+    print(f'\n---- Insert {len(set(val for dic in scanners for val in dic.values()))} entry(s) into experiment.Equipment ----')
+    Equipment.insert(scanners, skip_duplicates=True)
 
-    # ---------- Create ProcessingTask for each scan ----------
+    print(f'\n---- Insert {len(sessions)} entry(s) into experiment.Session ----')
+    Session.insert(sessions, skip_duplicates=True)
+    
+    print(f'\n---- Insert {len(session_directories)} entry(s) into experiment.Session.Directory ----')
+    Session.Directory.insert(session_directories, skip_duplicates=True)
 
-    # suite2p
-    imaging.ProcessingTask.insert([{**sc, 'paramset_idx': 0, 'task_mode': 'load'}
-                                   for sc in scan.Scan.fetch('KEY')], skip_duplicates=True)
+    print(f'\n---- Insert {len(scans)} entry(s) into scan.Scan ----')
+    scan.Scan.insert(scans, skip_duplicates=True)
 
-    # caiman - 2D
-    imaging.ProcessingTask.insert([{**sc, 'paramset_idx': 1, 'task_mode': 'load'}
-                                   for sc in (scan.Scan & (scan.ScanInfo & 'ndepths = 1')).fetch('KEY')],
-                                  skip_duplicates=True)
+    print(f'\n---- Insert {len(processing_tasks)} entry(s) into imaging.ProcessingTask ----')
+    imaging.ProcessingTask.insert(processing_tasks, skip_duplicates=True)
```

Comment:
we really should not need `skip_duplicates=True` here, if there's a duplication then that indicates bug in the code somewhere

Reply:
what about the case where the user appends to `subjects.csv` or `sessions.csv`, and reruns this script

Reply:
we should already have a safeguard against that, with the `if session_key not in Session.proj()`

Reply:
makes sense, thanks